lib/internal/modules/esm/assert.js

```diff
@@ -0,0 +1,104 @@
+'use strict';
+
+const {
+  ArrayPrototypeIncludes,
+  ArrayPrototypeFind,
+  ObjectCreate,
+  ObjectEntries,
+  ObjectValues,
+  ObjectPrototypeHasOwnProperty,
+  Symbol,
+} = primordials;
+
+const {
+  ERR_IMPORT_ASSERTION_TYPE_FAILED,
+  ERR_IMPORT_ASSERTION_TYPE_MISSING,
+  ERR_IMPORT_ASSERTION_TYPE_UNSUPPORTED,
+} = require('internal/errors').codes;
+
+const kImplicitAssertType = Symbol('implicit assert type');
+
+/**
+ * Define a map of module formats to import assertion types (the value of `type`
+ * in `assert { type: 'json' }`).
+ * @type {Map<string, string | typeof kImplicitAssertType}
+ */
+const formatTypeMap = {
+  '__proto__': null,
+  'builtin': kImplicitAssertType,
+  'commonjs': kImplicitAssertType,
+  'json': 'json',
+  'module': kImplicitAssertType,
+  'wasm': kImplicitAssertType, // Should probably be 'webassembly' per https://github.com/tc39/proposal-import-assertions
+};
+
+function findExpectedFormat(type) {
+  return ArrayPrototypeFind(
+    ObjectEntries(formatTypeMap),
```

Comment:
this entries array can be cached at module level, so it doesn't have to be recomputed each time

Reply:
This is the error path, should we really bother optimising it? It's only run when trying to import a JSON module without an assertion.

Reply:
I agree it doesn't matter much, but it's a really small and easy change. you could even lazily memoize it so that there's no memory usage prior to the error path.

Reply:
Turned out I didn't need that function at all, so I removed it. I've implemented caching for `Object.values` as you suggested below.